### PR TITLE
Start using 'new' subscribe_entities API instead of legacy state change

### PR DESCRIPTION
**Pull request by bgoncal**

### CODE REVIEW

**Comment by gloaysa** [comment]
hey, I know this is a draft, but still I wanted to test out your changes since I'm only interested on subscribing to specific entities.

I came up with this simple service to test:

```swift
import Foundation
import HAKit

struct HAService {
	let connection = HAKit.connection(configuration: .init(
		connectionInfo: {
			// Connection is required to be returned synchronously.
			// In a real implementation, handle both URL/connection info without crashing.
			try! .init(url: URL(string: "http://homeassistant.local:8123")!)
		},
		fetchAuthToken: { completion in
			// Access tokens are retrieved asynchronously, but be aware that Home Assistant
			// has a timeout of 10 seconds for sending your access token.
			completion(.success("my_long_lived_access_token"))
		}
	))
	
	func doSomething() {
		connection.send(.init(
			type: .subscribeEntities,
			data: ["entity_ids": ["todo.guille"]])
		) { result in
			print(result)
			switch result {
			case let .success(user):
				// e.g. user.id or user.name are available on the result object
				print(user)
			case let .failure(error):
				// an error occurred with the request
				print(error)
			}
		}
	}
}
```

and the response is always: `success(HAKit.HAData.empty)`.

I have tried this same message on Postman:

imagen

And it works (even when asking for 2 entities).

Is this what you were trying to implement? Based on your code I'd think so. Now I don't know if you have it on draft because you are not done with the whole implementation or, as I thought, because you have pending test to take care of.

I'd be happy to contribute to this PR if pointed to the right direction.

Thanks!

**Comment by bgoncal** [comment]
> hey, I know this is a draft, but still I wanted to test out your changes since I'm only interested on subscribing to specific entities.
> 
> I came up with this simple service to test:
> 
> ```swift
> import Foundation
> import HAKit
> 
> struct HAService {
> 	let connection = HAKit.connection(configuration: .init(
> 		connectionInfo: {
> 			// Connection is required to be returned synchronously.
> 			// In a real implementation, handle both URL/connection info without crashing.
> 			try! .init(url: URL(string: "http://homeassistant.local:8123")!)
> 		},
> 		fetchAuthToken: { completion in
> 			// Access tokens are retrieved asynchronously, but be aware that Home Assistant
> 			// has a timeout of 10 seconds for sending your access token.
> 			completion(.success("my_long_lived_access_token"))
> 		}
> 	))
> 	
> 	func doSomething() {
> 		connection.send(.init(
> 			type: .subscribeEntities,
> 			data: ["entity_ids": ["todo.guille"]])
> 		) { result in
> 			print(result)
> 			switch result {
> 			case let .success(user):
> 				// e.g. user.id or user.name are available on the result object
> 				print(user)
> 			case let .failure(error):
> 				// an error occurred with the request
> 				print(error)
> 			}
> 		}
> 	}
> }
> ```
> 
> and the response is always: `success(HAKit.HAData.empty)`.
> 
> I have tried this same message on Postman:
> 
> imagen
> 
> And it works (even when asking for 2 entities).
> 
> Is this what you were trying to implement? Based on your code I'd think so. Now I don't know if you have it on draft because you are not done with the whole implementation or, as I thought, because you have pending test to take care of.
> 
> I'd be happy to contribute to this PR if pointed to the right direction.
> 
> Thanks!

It is actually a work in progress, until the point I commited, I was able to get results when subscribed for entities. Feel free to contribute, I am busy at the moment so thats why I stopped in the middle.

Basically Im trying to get rid of the "get_states"

**Comment by bgoncal** [comment]
@gloaysa btw I just noticed in your sample that you are not subscribing for updates, I think what you are doing is to wait for a single response from the API right? so thats why it may not be returned to you

**Comment by bgoncal** [comment]
@gloaysa I just tested this, check it out:

`import Foundation
import HAKit

final class ContentViewModel: ObservableObject, HAConnectionDelegate {
    private var connection = HAKit.connection(configuration: .init(connectionInfo: {
        try! .init(url: URL(string: "https://<your-ha-server>/api/websocket")!)
    }, fetchAuthToken: { completion in
        completion(.success("<your-token>"))
    }), connectAutomatically: true)

    init() {
        connection.delegate = self
    }

    func connection(_ connection: HAConnection, didTransitionTo state: HAConnectionState) {
        print(state)
    }

    func doSomething() {
        connection.connect()
    }

    func subscribe() {
        print(connection.state)
        connection.subscribe(to: .subscribeEntities()) { cancellable, data in
            print(data)
        }
    }
}`

**Comment by gloaysa** [comment]
hey, sorry it took so long to come back to you, I haven't got the time to continue with this.

> I think what you are doing is to wait for a single response from the API right? so thats why it may not be returned to you

You are correct, I did notice it and tried subscribing, but I still couldn't subscribe to a single entity (the way you propose on https://github.com/home-assistant/HAKit/pull/46#issuecomment-1916489324 subscribes to all entities).

I've done some changes that allows me to do this now:

```swift
private func subscribeToHALists(_ listToSubscribe: [HAList]) -> Void {
		let entity_ids: [String] = listToSubscribe.compactMap { list in
			list.id
		}
		print("[HAService]: Subscribing to Lists: ", entity_ids)
		connection?.subscribe(
			to: .subscribeEntities(["entity_ids": entity_ids]),
			handler: {cancel, data in
				print(data)
			}
		)
	}
```

Should be easy enough to integrate in this PR or create a separate one to separate concerns.

My next step, once I have time, is to allow for calling services from within the library (https://github.com/home-assistant/core/pull/98610). Don't know when I'll be able to come back to this, tho.

**Comment by bgoncal** [comment]
> hey, sorry it took so long to come back to you, I haven't got the time to continue with this.
> 
> > I think what you are doing is to wait for a single response from the API right? so thats why it may not be returned to you
> 
> You are correct, I did notice it and tried subscribing, but I still couldn't subscribe to a single entity (the way you propose on [#46 (comment)](https://github.com/home-assistant/HAKit/pull/46#issuecomment-1916489324) subscribes to all entities).
> 
> I've done some changes that allows me to do this now:
> 
> ```swift
> private func subscribeToHALists(_ listToSubscribe: [HAList]) -> Void {
> 		let entity_ids: [String] = listToSubscribe.compactMap { list in
> 			list.id
> 		}
> 		print("[HAService]: Subscribing to Lists: ", entity_ids)
> 		connection?.subscribe(
> 			to: .subscribeEntities(["entity_ids": entity_ids]),
> 			handler: {cancel, data in
> 				print(data)
> 			}
> 		)
> 	}
> ```
> 
> Now I'm able to subscribe only to changes on the entities specified in `entity_ids` 😃
> 
> Should be easy enough to integrate in this PR or create a separate one to separate concerns.
> 
> My next step, once I have time, is to allow for calling services from within the library ([home-assistant/core#98610](https://github.com/home-assistant/core/pull/98610)). Don't know when I'll be able to come back to this, tho.

Better to create a separate PR for that to keep thing to the topic, this PR will focus specifically on removing the usage of "get_states" and start using the "new" entitites subscription model

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Changed: WebSocket connection will now enable compression.
 - Fixed: Calling `HAConnection.connect()` and `HAConnection.disconnect()` off the main thread no longer occasionally crashes.
 
+## [0.4] - 2024-02-01
```
Comment (zacwest) [inline CHANGELOG.md:15]: add to the unreleased above, probably

File: Source/Caches/HACache.swift [hunk 2]
```diff
@@ -70,6 +70,31 @@ public class HACache<ValueType> {
         )
     }
 
+    /// Create a cache that relies on subscription updates without initial population.
```
Comment (zacwest) [inline Source/Caches/HACache.swift:73]: i think it would be a good idea to break this out into a separate PR, mostly for writing tests distinct for it, since there's a lot in the subscription handling part below.

File: Source/Caches/HACache.swift [hunk 3]
```diff
@@ -381,7 +408,7 @@ public class HACache<ValueType> {
                 // No subscribers, do not connect.
                 return
             }
-            guard !state.isWaitingForPopulate else {
+            guard !state.isWaitingForPopulate || populateInfo == nil else {
```
Comment (zacwest) [inline Source/Caches/HACache.swift:411]: this feels like an invariant breakage that might be confusing later. suggest you do one of:

1. make startPopulate noop and do the normal code path for this case
2. set isWaitingForPopulate to false when starting the subscription for a populate-less subscription
3. make isWaitingForPopulate an enum with a case for not issuing populates, and don't change it below when in that case (e.g. `mutating func started()` changes the case around and checks)

File: Source/Caches/HACacheKeyStates.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            for (entityId, updates) in additions {
+                if let currentState = states[entityId] {
+                    if let updatedEntity = currentState.updatedEntity(compressedEntityState: updates) {
+                        states[entityId] = updatedEntity
+                    }
+                } else {
+                    do {
+                        states[entityId] = try updates.toEntity(entityId: entityId)
+                    } catch {
+                        HAGlobal.log(.error, "Failed adding new entity: \(error)")
```
Comment (zacwest) [inline Source/Caches/HACacheKeyStates.swift:36]: what should we do if an error occurs here? that means our local state is invalid, right? what possible errors can happen?

File: Source/Data/HAEntity+CompressedEntity.swift [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import Foundation
+
+extension HAEntity {
+    func updatedEntity(compressedEntityState: CompressedEntityState) -> HAEntity? {
```
Comment (zacwest) [inline Source/Data/HAEntity+CompressedEntity.swift:4]: why not make this mutating instead? i feel like it reads easier and since it's a struct there's no reason to worry about mutating underneath somebody using the old one

e.g., this feels good:

```
mutating func update(from state: CompressedEntityState)
mutating func add(_ state: CompressedEntityState)
mutating func subtract(_ state: CompressedEntityState)
```

File: Source/Data/HAEntity+CompressedEntity.swift [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            attributes: (compressedEntityState.attributes as? [String: Any]) ?? attributes.dictionary,
+            context: .init(id: compressedEntityState.context ?? "", userId: nil, parentId: nil)
+        )
+    }
+
+    func updatedEntity(adding compressedEntityState: CompressedEntityState) -> HAEntity? {
+        var newAttributes = attributes.dictionary
+        (compressedEntityState.attributes as? [String: Any])?.forEach({ key, value in
+            newAttributes[key] = value
+        })
```
Comment (zacwest) [inline Source/Data/HAEntity+CompressedEntity.swift:20]: the dance around attributes here makes me think you should probably make the HAEntityAttributes know about what you are trying to do, instead

File: Source/Data/HACompressedEntity.swift [hunk 7]
```diff
@@ -0,0 +1,75 @@
+import Foundation
+
+public struct CompressedStatesUpdates: HADataDecodable {
```
Comment (zacwest) [inline Source/Data/HACompressedEntity.swift:3]: i cannot really super well remember why i made the `HA` prefix for everything (i think i've been bit by namespace clashes too much) but it would be nice to preserve it

File: Source/Data/HACompressedEntity.swift [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    public let add: [String: CompressedEntityState]?
+    public let remove: [String]?
+    public let change: [String: CompressedEntityDiff]?
```
Comment (zacwest) [inline Source/Data/HACompressedEntity.swift:6]: no need to do `let` - this is a struct

File: Source/Data/HACompressedEntity.swift [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    public init(data: HAData) throws {
+        self.add = try? data.decode("a")
+        self.remove = try? data.decode("r")
+        self.change = try? data.decode("c")
+    }
+}
+
+public struct CompressedEntityState: HADataDecodable {
+    public let state: String
+    public let attributes: NSDictionary?
```
Comment (zacwest) [inline Source/Data/HACompressedEntity.swift:17]: why NSDictionary in swift?

File: Source/Data/HACompressedEntity.swift [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+    public init(data: HAData) throws {
+        self.state = try data.decode("s")
+        self.attributes = try? data.decode("a")
+        self.context = try? data.decode("c")
+        self.lastChanged = try? data.decode("lc")
+        self.lastUpdated = try? data.decode("lu")
+    }
+
+    public var lastChangedDate: Date? {
```
Comment (zacwest) [inline Source/Data/HACompressedEntity.swift:30]: see `HADecodeTransformable` - it already handles dates; please add support for Double-based dates as well and you won't need to do this

File: Source/Data/HACompressedEntity.swift [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+    public var lastUpdatedDate: Date? {
+        if let lastUpdated {
+            return Date(timeIntervalSince1970: lastUpdated)
+        } else {
+            return lastChangedDate
+        }
+    }
+
+    func toEntity(entityId: String) throws -> HAEntity {
```
Comment (zacwest) [inline Source/Data/HACompressedEntity.swift:46]: generally `asXYZ()` is what you find in swift nomenclature for this

File: Source/Data/HACompressedEntity.swift [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        let data = HAData(value: [
+            "entity_id": entityId,
+            "state": state,
+            "last_changed": lastChangedDate ?? Date(),
+            "last_updated": lastUpdatedDate ?? Date(),
+            "attributes": attributes as? [String: Any] ?? [:],
+            "context": ["id": context ?? ""],
+        ])
```
Comment (zacwest) [inline Source/Data/HACompressedEntity.swift:54]: why drop into a dictionary when you could initialize it directly?

**Review by zacwest — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (bgoncal) [reply CHANGELOG.md:15]: Can I add everything in the unreleased section to 0.4 and release it afterward?

File: Source/Caches/HACacheKeyStates.swift (hunk 4 above)
Comment (bgoncal) [reply Source/Caches/HACacheKeyStates.swift:36]: It means the updates coming from the subscription couldn't be converted into HAEntity, logging as error looks correct and then address the issue when/if happens

File: Source/Data/HACompressedEntity.swift (hunk 12 above)
Comment (bgoncal) [reply Source/Data/HACompressedEntity.swift:54]: I don't recall now why I did that but I think it was to reuse the logic of extracting domain from entityId, anyway, I will update it, thanks

File: Source/Data/HACompressedEntity.swift (hunk 9 above)
Comment (bgoncal) [reply Source/Data/HACompressedEntity.swift:17]: My bad

File: CHANGELOG.md (hunk 1 above)
Comment (zacwest) [reply CHANGELOG.md:15]: Yeah

File: Source/Caches/HACache.swift (hunk 2 above)
Comment (zacwest) [reply Source/Caches/HACache.swift:73]: Adding support for "populate-less" then using it in another PR

File: Source/Data/HACompressedEntity.swift (hunk 8 above)
Comment (zacwest) [reply Source/Data/HACompressedEntity.swift]: A few reasons: it makes writing tests easier in some contexts, it allows you to mutate data easier (modify a copy rather than initialize an entirely new one), and it's already a language feature (you can't change the value anywhere anyway)

File: Source/Caches/HACache.swift (hunk 3 above)
Comment (bgoncal) [reply Source/Caches/HACache.swift:411]: I think the way it is written now is correct, it reads like "If it is not waiting for populate or if there is no populate then move forward", I can add a comment if that helps.

Before doing that I tried number "2", but then I got those issues with reentrant access in cache

File: Source/Caches/HACache.swift (hunk 2 above)
Comment (bgoncal) [reply Source/Caches/HACache.swift:73]: Ok, here they are:
https://github.com/home-assistant/HAKit/pull/48
and 
https://github.com/home-assistant/HAKit/pull/49

I'll be closing this one

**Comment by bgoncal** [comment]
This PR was splitted in 2
https://github.com/home-assistant/HAKit/pull/48
and
https://github.com/home-assistant/HAKit/pull/49
